src/x11/util.c

```diff
@@ -163,24 +177,48 @@ adopt_atoms(struct xkb_context *ctx, xcb_connection_t *conn,
     xcb_get_atom_name_cookie_t cookies[SIZE];
     const size_t num_batches = ROUNDUP(count, SIZE) / SIZE;
 
+    if (!ctx->x11_atom_cache) {
+        ctx->x11_atom_cache = calloc(1, sizeof(struct x11_atom_cache));
+    }
+    /* Can be NULL in case the malloc failed. */
+    struct x11_atom_cache *cache = ctx->x11_atom_cache;
+    if (cache && cache->conn != conn) {
+        cache->conn = conn;
+        cache->len = 0;
+    }
+
+    memset(to, 0, count * sizeof(*to));
+
     /* Send and collect the atoms in batches of reasonable SIZE. */
     for (size_t batch = 0; batch < num_batches; batch++) {
         const size_t start = batch * SIZE;
         const size_t stop = MIN((batch + 1) * SIZE, count);
 
         /* Send. */
-        for (size_t i = start; i < stop; i++)
+        for (size_t i = start; i < stop; i++) {
+            if (cache) {
+                for (size_t c = 0; c < cache->len; c++) {
+                    if (cache->cache[c].from == from[i]) {
+                        to[i] = cache->cache[c].to;
+                        goto cache_hit;
+                    }
+                }
+            }
             if (from[i] != XCB_ATOM_NONE)
                 cookies[i % SIZE] = xcb_get_atom_name(conn, from[i]);
+            cache_hit: ;
```

Comment:
an indented label with a semicolon is a bit of a strange look tbh. How about something like this instead:

```c
   for (size_t i = start; i < stop; i++) {
       bool cache_hit = false;
       for (size_t c = 0; cache && c < cache->len; c++) {
           if (cache->cache[c].from == from[i]) {
               to[i] = cache->cache[c].to;
               cache_hit = true;
               break;
           }
       }
       if (!cache_hit && from[i] != XCB_ATOM_NONE)
           cookies[i % SIZE] = xcb_get_atom_name(conn, from[i]);
   }
```